### PR TITLE
Move StatefulSetAutoDeletePVC feature gate to beta

### PR DESCRIPTION
**Pull request by mattcary**
/kind feature

#### What this PR does / why we need it:

Move statefulset autodelete feature to beta.

```release-note
StatefulSetAutoDeletePVC feature gate promoted to beta.
```

- [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/1847-autoremove-statefulset-pvcs)

/sig apps

### CODE REVIEW

**Comment by mattcary** [comment]
conversation on the metrics is ongoing, but I want to get the review started on this.

/unhold
/retest

**Comment by mattcary** [comment]
@kow3ns @soltysh 

The apps reviewer on the metrics PR does not think the metrics should be added. I think that unblocks this PR. Can you approve? (or comment on the other PR). Thanks!

**Comment by Priyankasaggu11929** [comment]
Hello :wave:, 1.25 Enhancements Lead here.

Unfortunately, this enhancement did not meet the [code freeze](https://github.com/kubernetes/sig-release/blob/master/releases/release_phases.md#code-freeze) criteria because there are still unmerged k/k code PRs.

If you still wish to progress this enhancement in v1.25, please file an [exception request](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md). Thank you so much!

**Comment by janetkuo** [comment]
The change itself looks good. Presumably its blocker needs to be merged first: https://github.com/kubernetes/kube-state-metrics/pull/1876

**Review by janetkuo — APPROVED** [review]

File: pkg/features/kube_features.go
Comment (mattcary) [reply pkg/features/kube_features.go]: fixed

Comment (janetkuo) [reply pkg/features/kube_features.go]: FYI https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages says beta features are enabled by default.

**Review by janetkuo — APPROVED** [review]
/lgtm

**Comment by leonardpahlke** [comment]
Exception request submitted by @mattcary has been [accepted](https://groups.google.com/g/kubernetes-sig-release/c/pFlz9WF40BA/m/BKfn-1YiBAAJ) by the release team!

Your updated deadline to make any changes to your PR is [18:00 PST Thursday 10th November 2022](https://everytimezone.com/s/a62a5a3d).

/milestone v1.26

File: pkg/apis/apps/fuzzer/fuzzer.go
Comment (janetkuo) [inline pkg/apis/apps/fuzzer/fuzzer.go:56]: I looked at `types.go` and the api doc for `PersistentVolumeClaimRetentionPolicy` needs an update as well:

```go
// [...] This requires the
// StatefulSetAutoDeletePVC feature gate to be enabled, which is alpha.

PersistentVolumeClaimRetentionPolicy *StatefulSetPersistentVolumeClaimRetentionPolicy 
```

**Review by janetkuo — APPROVED** [review]
Just a nit, lgtm otherwise.

File: pkg/apis/apps/fuzzer/fuzzer.go
Comment (mattcary) [reply pkg/apis/apps/fuzzer/fuzzer.go:56]: ah, thanks for catching.

**Review by janetkuo — APPROVED** [review]
/lgtm

File: pkg/apis/apps/v1beta2/defaults_test.go
Comment (liggitt) [inline pkg/apis/apps/v1beta2/defaults_test.go:235]: it's really surprising the same test in pkg/apis/apps/v1/defaults_test.go didn't need updating

**Comment by liggitt** [comment]
/approve
for API fuzzer change and RBAC policy update

**Comment by liggitt** [comment]
/hold

looking at what this feature gate guards, it looks like UpdatePodClaimForRetentionPolicy updates a claim object retrieved from a lister without copying it first... this will break sharedinformer caches (and panic in tests with the mutation detector enabled)

**Comment by liggitt** [comment]
This gate protects code that modifies a *claim* (which is new... see the new `update` RBAC permission required), and the claim is obtained directly from a lister and modified without copying it

**Comment by mattcary** [comment]
> This gate protects code that modifies a _claim_ (which is new... see the new `update` RBAC permission required), and the claim is obtained directly from a lister and modified without copying it

Ah, you're absolutely right. Good catch.

**Comment by liggitt** [comment]
oh, the feature tag should be taken off the e2e so that it runs in presubmit

**Comment by mattcary** [comment]
Will update e2e tests and  delay this to the next release.

**Pull request closed without merging**
